### ISSUE TITLE
[Request] Relic's AIFC (.aifr)

### ISSUE THREAD

**Issue by gero-pardo** [vgmstream/vgmstream#907]
Relic ran a series of AIFF-Compressed variants during it's earliest days.
It consisted of a custom lossy compression algorithm for AIFF coded by Relic, maintained through most of the 2000s across various games.
The format was developed and renamed multiple times across the company's life, originally AIFR (AIF-Relic), then FDA (Fast Digital Audio), all sharing the same AIFC encoding as late as 2006's Company of Heroes. 

From my personal rips, I know of some formats:

.aifr - Homeworld and Homeworld: Cataclysm
First gen .fda (Homeworld 2) & Second gen .fda (Dawn of War, Company of Heroes & presumably Impossible Creatures) -- Already implemented for this project.

"Officially", the aifc encoder, decoder and an aif* Winamp plugin, all made by a Relic dev, surfaced in the early 2000s. 

I've tried to make the plugin myself, but I quickly realized my meagre Js programming expertise will not be enough to figure out how to decompile and analyse the decoder and compile a foobar2000 plugin in a mere winter vacation's time. I believe your reconstructed decoder should work, but I'm adding all relevant links in the strange case it doesn't:

- AIFC specs from Apple: http://www-mmsp.ece.mcgill.ca/Documents/AudioFormats/AIFF/AIFF.html
- Relic's AIFC custom encoder/decoder plus analysis of FDA & AIFR by a modder of the games: https://mega.nz/file/CqBDmQiA#_Jzd2MsjSozvcfedzkNiEYxErRW0E8J_3VJbB0NgJd0
- Relic's Winamp plugin: https://mega.nz/file/fjQXiIzI#NaO0LJIse87GSsrUSyS7Z12h0EUri7nXZK-uUxkZ5dk

Sample files :
- Homeworld .aifr (adagio for strings variant) - https://mega.nz/file/f6QVQYoZ#D5Ku5DeF5F1cl6EW5glNh4Ty43OKjHprSK5_0JAv4yI
- Homeworld Cataclysm .aifr (credit track) - https://mega.nz/file/LmZjCaJb#RSnGC664-242CtxSYuK2s3-bWNeUcl4dSmyVbw3DQeI

Off topic question: May I ask how you decompiled dec.exe?

**Comment by bnnm** [comment]
Those files aren't proper AIFC so they fail. Rename to .aif (or .aifc, .fda, etc), and add 4 bytes to the FORM and SSND sizes (32b value at 0x04 and 0x4c), then they'll play.

Homeworld / Cataclysm don't use .aifr but raw-ish data. Presumably the extraction tool took this data, pasted an AIFC header (with wrong sizes) and a new extension. Generally we don't add support for made-up extensions or wrong files if not found as-is in games.

Instead I made a quick python script to fix your files: [aifr-fixer.zip](https://github.com/vgmstream/vgmstream/files/6911007/aifr-fixer.zip) (makes separate .fda)

dec.exe was decompiled using IDA and standard tricks (unpacking UPX first). If you are interested here is the dec.c file I made before adding to vgmstream: https://gist.github.com/bnnm/bf6d4d3cb4335ff1d09ff46481eea2f4

**Comment by gero-pardo** [comment]
>Homeworld / Cataclysm don't use .aifr but raw-ish data. Presumably the extraction tool took this data, pasted an AIFC header (with wrong sizes) and a new extension. Generally we don't add support for made-up extensions or wrong files if not found as-is in games.

Ah, my apologies! I extracted using tools made by the fanbase (a tool named Penesound), didn't occur to me they were jury-rigging files. 
Raw-ish data? You mean it's like an uncut audio rip? Or if I were to extract the original container files (.bnk, .wxd and vce) I could obtain the original file format?

**Comment by bnnm** [comment]
Original data in .wxd/vce has no AIFC header, it's just plain Relic audio data (what comes after "SSND" in .aif/fda) pasted together one after another. Tools separate that and put a new AIFC in front, but data is unchanged.

An AIFC header contains audio info (mono/stereo, sample rate, etc), but game just hardcodes this (wxd=stereo, vce=mono, etc) so has no need for a header. Later games do use a standard AIFC or [FDA](https://github.com/vgmstream/vgmstream/blob/master/src/meta/fda.c) though.

The AIFC is needed for external players since they can't figure out format otherwise, there is no difference in data or audio. Other players aren't that picky and will accept not-quite proper, made-up AIFC, but vgmstream wants a bit more exact files to avoid bad rips, so the python script only fixes the wonky AIFC header.

**Comment by gero-pardo** [comment]
>Homeworld / Cataclysm don't use .aifr but raw-ish data. Presumably the extraction tool took this data, pasted an AIFC header (with wrong sizes) and a new extension. Generally we don't add support for made-up extensions or wrong files if not found as-is in games.

Ah, my apologies! I extracted using tools made by the fanbase (a tool named Penesound), didn't occur to me they were jury-rigging files. 
Raw-ish data? The original files were in a number of containers (.bnk, .wxd and vce), could those be decompressed/decompiled to find the actual original formats inside, or they are like an audio rip - that requires a .cue like file to read each audio segment?

**Comment by bnnm** [comment]
Oh right, there is a CUE-like .wxh header file, but inside the main .big (can be extracted with tools like BigViewer.exe). It has the offsets/info the engine uses to play data in the .wxd. I could add this format if you have both .wxd+.wxh in the same dir, that would be the closest to keeping it untouched.
For voices I think they use various .lut instead that are a bit different.

**Comment by gero-pardo** [comment]
>I could add this format if you have both .wxd+.wxh in the same dir, that would be the closest to keeping it untouched.

Right, here's both games' WXD, VCE and the soundfx folder with the BNK sound effects containers and their associated LUT and WXH files.
It seems multiple LUTs call different sections of the VCE and BNK files. Does this complicate matters?

HM: https://mega.nz/file/SrBXRAYa#SqDF3DlbX_rZog08IiPGOzRrmSOwBrFEb5oUPWbMvMA
HM Cataclysm: https://mega.nz/file/3rAVCaiS#iTFIjHRvfNqyb_n5CLjH8LJ3V2yz61LqZ7Fsy6UQPdo

**Comment by bnnm** [comment]
Can you post some of the .aifr made by your tool from bnk/vce? I'm having trouble to decode them so an alt file for comparison would probably help.

**Comment by gero-pardo** [comment]
I can give you the tool itself: https://subversion.assembla.com/svn/homeworld-and-homeworld-2-tool-box/HW1%20-%20Penesound/
Here's a collection of samples: https://mega.nz/file/ruwUCJAD#pTz1dI2S4CNMw4O23LXhBx3CCrxbvYgU_J8xFbvqwvg
wrote a small readme with details and some notes. Comments in this vid claim hardcoded filtering for each clip, but I can't confirm it's verasity: https://www.youtube.com/watch?v=nfClyVeC-S

**Comment by bnnm** [comment]
Added .whx+wxh and .bnk, I've yet to think to handle .vce

**Comment by gero-pardo** [comment]
What's so particular about them?

**Comment by bnnm** [comment]
The *sentence.lut seem to have header/offsets, but there are a lot of dummy entries and don't follow .vce order, plus may be linked to *phrase.lut somehow. Need to think how to make them usable or if there are better ways to get offsets.

**Comment by bnnm** [comment]
Try selecting playlist > shift + right click > Tagging > Reload info from file(s), also see [here](https://github.com/vgmstream/vgmstream/blob/master/README.md#foo_input_vgmstream-foobar2000-plugin)

**Comment by gero-pardo** [comment]
Understood. Thanks. Should I close the issue?

**Comment by bnnm** [comment]
If you still want .vce support you can leave it open, it may take a while tho.
